### PR TITLE
Add: First and Follow

### PR DESCRIPTION
**Pull request by noob77777**
# First and Follow

## Description
This pr adds first and follow computation for context-free-grammar productions.
Continuation of #10.

closes #10 

## Remaining tasks

- [x] Add unit tests
- [x] Add comments and clean up code a bit
- [x] Remove iterator based loops (Optional)

## Further work
After unit tests have been written check the recursive functions for scope of optimization.

### CODE REVIEW

File: src/grammar/grammar.cpp [hunk 1]
```diff
@@ -1,11 +1,30 @@
...
+grammar::Rules GetRulesForParent(const grammar::Productions &productions, const std::string &parent) {
+  grammar::Rules rules;
+  for (const auto &production : productions) {
+    if (production.GetParent() == parent) {
+      for (const auto &rule : production.GetRules()) {
+        rules.push_back(rule);
+      }
+      break;
+    }
+  }
```
Comment (noob77777) [inline src/grammar/grammar.cpp:20]: newline

File: src/utils/first_follow.cpp [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+      firsts.insert(make_pair(key, std::vector<std::string>(0)));
+      finished = false;
+    }
+
+    for (const auto &rules : GetRulesForParent(augmented_grammar, key)) {
+      std::vector<std::string>::const_iterator symbol_itr;
+      for (symbol_itr = rules.GetEntities().begin(); symbol_itr != rules.GetEntities().end(); symbol_itr++) {
+        std::vector<std::string> first = calc_recursive(*symbol_itr, path);
+        for (const auto &der : first) {
+          if (der != "E" && find(firsts[key].begin(), firsts[key].end(), der) == firsts[key].end()) {
```
Comment (noob77777) [inline src/utils/first_follow.cpp]: unreadable code
random constants
wtf is der?

File: src/utils/first_follow.cpp [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+            }
+          }
+        }
+      }
+    }
+  };
+
+  while (!finished) {
+    finished = true;
+    calc_recursive("S", std::vector<std::string>(0));
```
Comment (noob77777) [inline src/utils/first_follow.cpp]: random constant "S"

File: src/utils/first_follow.cpp [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+
+std::unordered_map<std::string, std::vector<std::string>> CalcFollows(
+    const grammar::Productions &augmented_grammar, const std::unordered_map<std::string, bool> nullables,
+    const std::string &start_symbol) {
+  std::unordered_map<std::string, std::vector<std::string>> follows;
+  std::unordered_map<std::string, std::vector<std::string>> firsts = CalcFirsts(augmented_grammar, nullables);
+  bool finished = false;
+  for (const auto &production : augmented_grammar) {
+    if (production.GetParent() == start_symbol) {
+      follows.insert(make_pair(production.GetParent(), std::vector<std::string>(1, "$")));
```
Comment (noob77777) [inline src/utils/first_follow.cpp]: create global constant for "$"

**Comment by noob77777** [comment]
Ignore changes to grammar_test files.
Changed naming from
2, 3, 3b, 4, 5 ... to 2, 3, 4, 5, 6 ...

**Review by Aritra-Samanta — COMMENTED** [review]
Noice observation

File: src/grammar/grammar.cpp [hunk 5]
```diff
@@ -1,11 +1,32 @@
...
+      for (const auto &rule : production.GetRules()) {
+        rules.push_back(rule);
+      }
+
+      break;
+    }
+  }
+
+  return rules;
+}
```
Comment (bisakhmondal) [inline src/grammar/grammar.cpp]: Redundant Codes
```suggestion
grammar::Rules GetRulesForParent(const grammar::Productions &productions, const std::string &parent) {
  for (const auto &production : productions) {
    if (production.GetParent() == parent) {
      return  production.GetRules();
    }
  }
  return grammar::Rules();
  }
```

File: codecov.yml [hunk 6]
```diff
@@ -12,7 +12,7 @@ coverage:
         threshold: 5%
     patch:
       default:
-        target: 80%
+        target: 90%
```
Comment (TheSYNcoder) [inline codecov.yml:15]: Okaaaay

File: src/include/grammar/grammar.h [hunk 7]
```diff
@@ -60,6 +60,19 @@ class Production {
 
 using Productions = std::vector<Production>;
 
+/**
+ * Search if a production exists for a given parent
+ * utility function
+ * @return if parent symbol is present in productions
+ */
+bool SearchParent(const grammar::Productions & /*productions*/, const std::string & /*parent*/);
+
```
Comment (bisakhmondal) [inline src/include/grammar/grammar.h]: `CheckParent` os  `IsParentExists` or `HasParent` would be nice

File: src/include/utils/first_follow.h [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+ * @returns an unordered_map keyed by each symbol in the grammar of booleans
+ * true if symbol is nullable, false otherwise
+ */
+std::unordered_map<std::string, bool> CalcNullables(const grammar::Productions & /*augmented_grammar*/);
+
+/**
+ * For each non terminal in given set of productions computes Firsts
+ * @returns an unordered_map keyed by non terminals in the grammar of a vector of terminals
+ */
+std::unordered_map<std::string, std::vector<std::string>> CalcFirsts(
```
Comment (bisakhmondal) [inline src/include/utils/first_follow.h]: How about `std::unordered_map<std::string, std::vector<std::string>>` ==> `ListMap` or `SymbolsMap`
Large data structure definition, reduces readability, increase redundancy

File: src/utils/first_follow.cpp [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+          nullables[symbol] = false;
+        }
+      }
+    }
+  }
+
+  // EPSILON is nullable by defination
+  if (nullables.find(std::string(grammar::EPSILON)) != nullables.end()) {
+    nullables[std::string(grammar::EPSILON)] = true;
+  }
```
Comment (bisakhmondal) [inline src/utils/first_follow.cpp]: Why needs checking, if it is nullable by default?

File: src/utils/first_follow.cpp [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+
+  // recursively compute for all non terminals
+  std::function<bool(const std::string &, std::vector<std::string> &)> calc_recursive;
+  // lambda function that returns boolean value
+  calc_recursive = [&](const std::string &key, std::vector<std::string> &path) {
+    if (find(path.begin(), path.end(), key) != path.end()) {
+      return false;
+    }
```
Comment (bisakhmondal) [inline src/utils/first_follow.cpp:31]: please maintain consistency,
`path.find(....)`

File: src/include/utils/first_follow.h [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    const grammar::Productions & /*augmented_grammar*/, const std::unordered_map<std::string, bool> & /*nullables*/);
+
+/**
+ * For each non terminal in given set of productions computes Follows
+ * @returns an unordered_map keyed by non terminals in the grammar of a vector of terminals
+ */
+std::unordered_map<std::string, std::vector<std::string>> CalcFollows(
+    const grammar::Productions & /*augmented_grammar*/, const std::unordered_map<std::string, bool> & /*nullables*/,
+    const std::string & /*start_symbol*/);
+
```
Comment (TheSYNcoder) [inline src/include/utils/first_follow.h]: Clang-Tidy
readability-named-parameter

Why is it commented out ??

Comment (noob77777) [reply src/include/utils/first_follow.h]: This is allowed.

File: src/utils/first_follow.cpp [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+
+    for (const auto &rule : GetRulesForParent(augmented_grammar, key)) {
+      std::vector<std::string>::const_iterator symbol_itr;
+      for (symbol_itr = rule.GetEntities().begin(); symbol_itr != rule.GetEntities().end(); symbol_itr++) {
+        // exit loop if non-nullable symbol is encountered
+        if (!calc_recursive(*symbol_itr, path)) {
+          break;
+        }
+      }
```
Comment (bisakhmondal) [inline src/utils/first_follow.cpp]: if possible, use range based loop

File: src/utils/first_follow.cpp [hunk 13]
```diff
@@ -0,0 +1,267 @@
...
+    }
+
+    // by default, a symbol is considered non-nullable
+    nullables[key] = false;
+    path.pop_back();
+    return false;
+  };
+
+  for (const auto &production : augmented_grammar) {
+    auto path = std::vector<std::string>(0);
```
Comment (bisakhmondal) [inline src/utils/first_follow.cpp]: any particular reason for creating a zero-length vector?

File: src/utils/first_follow.cpp [hunk 14]
```diff
@@ -0,0 +1,267 @@
...
+  // finished -> used to check if any new symbols are added to any FIRST(non-terminal) in a particular iteration
+  // useful in case of cycles in productions
+  bool finished = false;
+  for (const auto &production : augmented_grammar) {
+    for (const auto &rules : production.GetRules()) {
+      for (const auto &symbol : rules.GetEntities()) {
+        if (!SearchParent(augmented_grammar, symbol)) {
+          // if symbol is not present as a parent in grammar, then it is a terminal
+          // if X is a terminal, then FIRST(X) = {X}
+          firsts.insert(make_pair(symbol, std::vector<std::string>(1, symbol)));
```
Comment (bisakhmondal) [inline src/utils/first_follow.cpp]: maintain consistency

File: src/utils/first_follow.cpp [hunk 15]
```diff
@@ -0,0 +1,267 @@
...
+  calc_recursive = [&](const std::string &key, std::vector<std::string> &path) {
+    // FIRST(terminal) = {terminal}
+    if (!SearchParent(augmented_grammar, key)) {
+      return firsts[key];
+    }
+
+    // if key "belongs to" path, then first of key has already been calculated / visited
+    if (find(path.begin(), path.end(), key) != path.end()) {
+      return firsts[key];
+    }
```
Comment (bisakhmondal) [inline src/utils/first_follow.cpp:106]: same

File: src/utils/first_follow.cpp [hunk 16]
```diff
@@ -0,0 +1,267 @@
...
+
+    // marking current symbol as visited
+    path.push_back(key);
+
+    /* now only un-visited non-terminals are left to be explored */
+
+    // registering non-terminal in FIRST map if encountered for the first time
+    if (firsts.find(key) == firsts.end()) {
+      firsts.insert(make_pair(key, std::vector<std::string>(0)));
```
Comment (bisakhmondal) [inline src/utils/first_follow.cpp]: same, why be specific about zero-length vector when that comes by default?

File: src/utils/first_follow.cpp [hunk 17]
```diff
@@ -0,0 +1,267 @@
...
+    for (const auto &rules : GetRulesForParent(augmented_grammar, key)) {
+      // symbol_itr -> iterates through all symbols in the body of a production
+      std::vector<std::string>::const_iterator symbol_itr;
+      for (symbol_itr = rules.GetEntities().begin(); symbol_itr != rules.GetEntities().end(); symbol_itr++) {
+        // deriving FIRST(current symbol)
+        std::vector<std::string> first = calc_recursive(*symbol_itr, path);
+        // der -> each terminal or EPSILON in FIRST(current symbol)
+        for (const auto &der : first) {
+          if (der != std::string(grammar::EPSILON) &&
+              find(firsts[key].begin(), firsts[key].end(), der) == firsts[key].end()) {
```
Comment (bisakhmondal) [inline src/utils/first_follow.cpp:131]: same

File: src/utils/first_follow.cpp (hunk 10 above)
Comment (noob77777) [reply src/utils/first_follow.cpp:31]: `std::vector<std::string>.find(std::string)` doesn't exist.

**Review by bisakhmondal — CHANGES_REQUESTED** [review]

File: src/utils/first_follow.cpp (hunk 10 above)
Comment (bisakhmondal) [reply src/utils/first_follow.cpp:31]: really!! Oh GOD!!

Comment (bisakhmondal) [reply src/utils/first_follow.cpp:31]: Then we are good😘

File: src/utils/first_follow.cpp [hunk 18]
```diff
@@ -0,0 +1,267 @@
...
+        calc_recursive(production.GetParent());
+      }
+    }
+  }
+
+  // sorting FOLLOW mappings for convenience in presentation
+  for (const auto &production : augmented_grammar) {
+    sort(follows[production.GetParent()].begin(), follows[production.GetParent()].end());
+  }
+
```
Comment (TheSYNcoder) [inline src/utils/first_follow.cpp]: This step is redundant and will add extra to the compilation time.

File: src/utils/first_follow.cpp (hunk 9 above)
Comment (bisakhmondal) [reply src/utils/first_follow.cpp]: also typo🤡

File: src/utils/first_follow.cpp (hunk 18 above)
Comment (noob77777) [reply src/utils/first_follow.cpp]: This makes easier to write tests.

Comment (TheSYNcoder) [reply src/utils/first_follow.cpp]: That is not a valid reason.

File: codecov.yml (hunk 6 above)
Comment (bisakhmondal) [reply codecov.yml:15]: Make it 95🤡

File: src/utils/first_follow.cpp (hunk 14 above)
Comment (noob77777) [reply src/utils/first_follow.cpp]: what consistency?

Comment (noob77777) [reply src/utils/first_follow.cpp]: ?

Comment (TheSYNcoder) [reply src/utils/first_follow.cpp]: Please don't use insert in case of `unordered_map` , becomes really unreadable.

Comment (TheSYNcoder) [reply src/utils/first_follow.cpp:88]: What does `std::vector<std::string>(1, symbol))` mean here.

Comment (TheSYNcoder) [reply src/utils/first_follow.cpp]: A better way would be 
```
firsts[symbol] = {symbol};
```

File: src/utils/first_follow.cpp (hunk 16 above)
Comment (TheSYNcoder) [reply src/utils/first_follow.cpp]:
```
firsts[key] = {};
```

File: src/utils/first_follow.cpp (hunk 14 above)
Comment (bisakhmondal) [reply src/utils/first_follow.cpp]: exactly

File: src/utils/first_follow.cpp [hunk 19]
```diff
@@ -0,0 +1,267 @@
...
+    }
+
+    /* if X is a non-terminal and X -> Y1 Y2 Y3 ... Yk is a production, then
+       for 1 <= i <= k, FIRST(X) includes FIRST(Yi) iff FIRST(Y1 ... Yi-1) => EPSILON
+       and FIRST(Y1 ... Yi-1) => EPSILON iff for all 1 <= j <= i - 1, Yj => EPSILON */
+    for (const auto &rules : GetRulesForParent(augmented_grammar, key)) {
+      // symbol_itr -> iterates through all symbols in the body of a production
+      std::vector<std::string>::const_iterator symbol_itr;
+      for (symbol_itr = rules.GetEntities().begin(); symbol_itr != rules.GetEntities().end(); symbol_itr++) {
+        // deriving FIRST(current symbol)
```
Comment (TheSYNcoder) [inline src/utils/first_follow.cpp]: Suggestion
Change to 

```
for ( auto symbol & : rules.GetEntities() ) {
        ... 
}
```

File: src/utils/first_follow.cpp [hunk 20]
```diff
@@ -0,0 +1,251 @@
...
+    // registering non-terminal in FOLLOW map if encountered for the first time
+    if (nullables.find(key) != nullables.end()) {
+      path.pop_back();
+      return nullables[key];
+    }
+
+    for (const auto &rule : GetRulesForParent(augmented_grammar, key)) {
+      std::vector<std::string>::const_iterator symbol_itr;
+      for (symbol_itr = rule.GetEntities().begin(); symbol_itr != rule.GetEntities().end(); symbol_itr++) {
+        // exit loop if non-nullable symbol is encountered
```
Comment (TheSYNcoder) [inline src/utils/first_follow.cpp]: Change the loop structure please 🙏

File: src/include/utils/first_follow.h [hunk 21]
```diff
@@ -21,20 +21,22 @@ const char STRING_ENDMARKER[] = "$";
  */
 std::unordered_map<std::string, bool> CalcNullables(const grammar::Productions & /*augmented_grammar*/);
 
+using SymbolsMap = std::unordered_map<std::string, std::vector<std::string>>;
```
Comment (bisakhmondal) [inline src/include/utils/first_follow.h:24]: Noice. Thanks

**Review by bisakhmondal — DISMISSED** [review]
LGTM :+1:

File: src/include/utils/first_follow.h [hunk 22]
```diff
@@ -24,7 +24,8 @@ std::unordered_map<std::string, bool> CalcNullables(const grammar::Productions &
 using SymbolsMap = std::unordered_map<std::string, std::vector<std::string>>;
 
 /**
- * For each non terminal in given set of productions computes Firsts
+ * For each non terminal in given set of productions computes Firsts, and in case of terminals
+ * it returns the same.
  * @returns an unordered_map keyed by non terminals in the grammar of a vector of terminals
```
Comment (noob77777) [inline src/include/utils/first_follow.h]: ughh... @returns line

**Review by bisakhmondal — APPROVED** [review]

**Pull request merged**